azkaban-exec-server/src/main/java/azkaban/execapp/JobRunner.java

```diff
@@ -741,23 +784,13 @@ private Status prepareJob() throws RuntimeException {
         this.props.put(AbstractProcessJob.WORKING_DIR, this.workingDir.getAbsolutePath());
       }
 
-      if (this.props.containsKey(JobProperties.USER_TO_PROXY)) {
-        final String jobProxyUser = this.props.getString(JobProperties.USER_TO_PROXY);
-        if (this.proxyUsers != null && !this.proxyUsers.contains(jobProxyUser)) {
-          final String permissionsPageURL = getProjectPermissionsURL();
-          this.logger.error("User " + jobProxyUser
-              + " has no permission to execute this job " + this.jobId + "!"
-              + " If you want to execute this flow as " + jobProxyUser
-              + ", please add it to Proxy Users under project permissions page: " +
-              permissionsPageURL);
-          return null;
-        }
-      } else {
-        final String submitUser = this.getNode().getExecutableFlow().getSubmitUser();
-        this.props.put(JobProperties.USER_TO_PROXY, submitUser);
-        this.logger.info("user.to.proxy property was not set, defaulting to submit user " +
-            submitUser);
+      Optional<String> effectiveUserOptional = determineEffectiveUser();
+      // Fail if the effective user is not present
+      if (!effectiveUserOptional.isPresent()) {
+        return null;
```

Comment:
It seems strange to me to do so much work with Optionals only to return null anyway :).

Reply:
+1

Reply:
Thanks for this. The usage of optional instead of null is to have safety again the possible NPE. 
However, this method prepareJob() is either returned null or as Status.RUNNING. Hence I have changed it to have boolean as return type, and updated the place where it is called.